R/telemetry.R

```diff
@@ -715,18 +721,52 @@ Telemetry <- R6::R6Class( # nolint object_name_linter
         )
       }
     },
-
+    extract_cookie = function(cookie_string, cookie_name = "shiny_user_cookie") {
+      checkmate::assert_string(cookie_string, null.ok = TRUE)
+      checkmate::assert_string(cookie_name, null.ok = FALSE)
+      if (is.null(cookie_string)) return(NULL)
+      cookies <- strsplit(cookie_string, ";")[[1]]
+      cookies <- trimws(cookies)
+      for (cookie in cookies) {
+        parts <- strsplit(cookie, "=")[[1]]
+        if (length(parts) == 2 && parts[1] == cookie_name) {
+          # Check if the value looks like a SHA256 hash
+          cookie_value <- parts[2]
+          if (grepl("^[a-f0-9]{64}$", cookie_value)) {
+            return(cookie_value)
+          }
+        }
+      }
+      NULL
+    },
     get_user = function(
       session = shiny::getDefaultReactiveDomain(),
-      force_username = NULL
+      force_username = NULL,
+      track_anonymous_user = TRUE
     ) {
       if (!is.null(force_username)) return(force_username)
       if (isFALSE(is.null(session)) && isFALSE(is.null(session$user))) {
-        return(session$user) # POSIT Connect
+        session$user # POSIT Connect
       } else if (nzchar(Sys.getenv("SHINYPROXY_USERNAME"))) {
-        return(Sys.getenv("SHINYPROXY_USERNAME"))
+        Sys.getenv("SHINYPROXY_USERNAME")
+      } else if (track_anonymous_user) {
+        cookie_value <- private$extract_cookie(cookie_string = session$request$HTTP_COOKIE)
+        # cookie_value will be NULL if either not found or not generated using SHA256 algorithm.
+        if (is.null(cookie_value)) {
+          cookie_string <- paste0(
+            session$request$HTTP_USER_AGENT,
+            session$request$REMOTE_ADDR, Sys.time()
+          )
+          cookie_value <- digest::digest(cookie_string, algo = "sha256")
```

Comment:
Small suggestion to avoid unnecessary variable and `paste0` call

Also using `session$token` to avoid the unlikely case of 2 users with same browsers and with same IP (VPN or behind same organization) access it.

```suggestion
          cookie_value <- digest::digest(
            c(
              session$token,
              session$request$HTTP_USER_AGENT,
              session$request$REMOTE_ADDR,
              Sys.time()
            ),
            algo = "sha256"
          )
```

Reply:
Great Suggestion.
 Updated and tested the newly generated cookie value